specs/withdrawals.md

```diff
@@ -0,0 +1,176 @@
+  - [Rationale](#rationale)
+- [The L2 Withdrawal Contract](#the-l2-withdrawal-contract)
+- [Trie Interface](#trie-interface)
+- [Key differences from Current System](#key-differences-from-current-system)
+- [Security Considerations](#security-considerations)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+Withdrawals are transactions initiated on L2, and executed on L1. They may be used to transfer data, ether or both from
+L1 to L2.
+
+**Vocabulary note**:
+
+- _withdrawing transaction_ refers specifically to an L2 transaction
+- _withdrawn transaction_ refers specifically to an L1 transaction
+- _withdrawal_ can refer to the transaction at various stages
+
+## Constants
+
+| Name                  | Value     | Unit    |
+| --------------------- | --------- | ------- |
+| `FINALIZATION_WINDOW` | `604_800` | seconds |
+
+This `FINALIZATION_WINDOW` value is equivalent to 7 days.
+
+## Data Structures and Type Aliases
+
+1. A `WithdrawalMessage` is encoded in a struct as follows:
+
+    ```js
+    struct WithdrawalMessage {
+        uint256 nonce;
+        address sender;
+        uint256 value;
+        bytes message;
+    }
+    ```
+
+1. The `L2OutputTimestamp` is an alias for `uint256`, and MUST be a multiple of the `SUBMISSION_INTERVAL` described
+  in the [L2 Output](./proposals.md#constants) document.
+
+1. The `WithdrawalsRoot` is an alias for a `bytes32` value, corresponding to the [MPT][g-mpt]
+  storage root of the Withdrawals predeploy contract at `0x4200000000000000000000000000000000000015` (described below).
+
+1. The `WithdrawalsRootInclusionProof` proof contains the data necessary to prove that the provided `WithdrawalsRoot` is
+  included in the [SSZ merkleization](https://github.com/ethereum/consensus-specs/blob/dev/ssz/simple-serialize.md#merkleization)
+  of the `L2Output` as defined in the
+  [L2 output commitment construction](./proposals.md#l2-output-commitment-construction).
+
+1. A `WithdrawalMessageInclusionProof` is an MPT proof encoded in a struct as follows:
+
+   ```js
+   struct WithdrawalMessageInclusionProof {
+       WithdrawalMessage message;
+       bytes32 l2withdrawalsRoot;
+       bytes memory _key;      // storage key of the withdrawal message commitment
+       bytes memory _value;    // Always bytes32(1) (boolean validity status of the message)
+       bytes memory _proof;    // MPT inclusion proof for the key/value
+   }
+   ```
+1. This document also refers to the `L2Output` type as defined in the
+  [L2 output commitment construction](./proposals.md#l2-output-commitment-construction).
+
+## Withdrawal verification
+
+In order to verify a `WithdrawalMessage`, the following inputs must be provided:
+
+| Type                              | Name               |
+| --------------------------------- | ------------------ |
+| `WithdrawalMessage`               | `message`          |
+| `L2OutputTimestamp`               | `timestamp`        |
```

Comment:
That makes sense, but just so I'm clear, what is the intent with the timestamp used as the key in this mapping? 

https://gist.github.com/karlfloersch/02ce7938000680ce7820f39ba06bd9b1#file-staterootoracle-sol-L45

Reply:
Ahhhh! My goodness I'm sorry @maurelian -- I forgot that I implemented this to be super simple instead of building in features that will eventually be needed in the final version. For now we should just use the `timestamp` inside the output because it's easier, but in the long term we will definitely need to move to the model that I outlined in my comment. The reason why it's fine to use the output timestamp for now is because it doesn't effect the (already quite strong) trust assumptions around the sequencer who's submitting outputs.